README.md
# genie-dat-utils

read and write age of empires .dat files in nodejs.
send issues if there are any requests this package aims to be the most portable .dat file opener for AoE games.
currently fully supports AoE DE2 since 2023 Marc. 

[Install](#install) - [Usage](#usage) - [Supported Formats](#supported-formats) -
[Limitations](#limitations) - [API](#api) - [License: LGPL-3](#license)

[![npm][npm-image]][npm-url]
[![travis][travis-image]][travis-url]
[![standard][standard-image]][standard-url]

[npm-image]: https://img.shields.io/npm/v/genie-dat.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/genie-dat
[travis-image]: https://img.shields.io/travis/com/genie-js/genie-dat.svg?style=flat-square
[travis-url]: https://travis-ci.com/genie-js/genie-dat
[standard-image]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat-square
[standard-url]: http://npm.im/standard

## Install

```
npm install genie-dat-utils
```

## Usage

```js
var genieDat = require('genie-dat-utils')
var buffer = require('fs').readFileSync('/path/to/empires2_x1_p1.dat')
genieDat.load(buffer, (err, dat) => {
  console.log(dat.civilizations.map(civ => civ.name))
})
```

## Supported Formats

The data file format differs slightly between game versions. Versions supported by this library are:

 - `aoc` - All AoC-style formats, such as `empires2_x1.dat` of AoC 1.0 and `empires2_x1_p1.dat` of AoC 1.0c (the most common version). UserPatch and UserPatch based mods like WololoKingdoms and Portuguese Civ Mod also use this format. HD Edition versions before 5.0 also use this format.
 - `african-kingdoms` - The release of African Kingdoms changed the file format slightly; HD Edition versions 5.0 and up use this format. Note that this is used regardless of whether the African Kingdoms DLC is actually installed.

## Limitations

Age of Kings, SWGB, and the AoE1 and AoE:DE files are not yet supported. The goal is to expand support for these formats in the future.

Still write/save functionality is not done, but can be easily implemented using the existing codes.

## API

`genieDat.load(buffer, opts={}, cb)`

Decompress and load `buffer`, eg. a raw .dat file.

`opts` can be an object with properties:

 - `version` - A version string listed under [Supported Formats](#supported-formats). Defaults to `aoede2`.

`cb` is a Node-style callback receiving `(err, dat)`.
`dat` is a plain object representing the dat file contents. `console.log()` it to find out what's in it. There is a lot of junk, some of the more useful properties are:

 - `playerColors` - Player colours, mostly offsets into the main [palette](https://github.com/genie-js/jascpal) file
 - `techs` - Technology effects
 - `terrains` - Lists terrains
 - `civilizations` - Lists available civilizations
   - `civilizations[i].objects` - Lists unit statistics for each civ
 - `researches` - Lists available researches—this refers to the technology effects a lot

### `genieDat.loadRaw(buffer, opts={}, cb)`

Load an already decompressed `buffer`. Use this if you manually did `zlib.inflateRaw` or got an uncompressed buffer through some other means. Otherwise it is identical to `genieDat.load`.

## This project is forked from

The original [genie-dat was forked by mrommel](https://github.com/mrommel/genie-dat#supported-formats), who made a lot of improvements, I could also fork that repo, but it would be confusing. Also for visibility purposes and to finally group efforts in portability of genie dat files, I propose this new github repository.

## Support

If newer version comes out from AoEDE2 we just copy the protocol from the official: [genieutils](https://github.com/Tapsa/genieutils)

## License

This project is based heavily on [genieutils](https://github.com/Tapsa/genieutils) by apreiml and Tapsa, and on the [openage](https://github.com/sfttech/openage) [conversion script](https://github.com/sfttech/openage/tree/master/openage/convert/gamedata).

[LGPL-3](LICENSE)


CHANGELOG.md
# genie-dat-utils originally: genie-dat change log

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/).

## 0.5.0
* AoE DE2 support.
* Code improvements by adding better version supports.
* Rename some fields.

## 0.4.0
* Add root tech tree reading.
* Rename some fields.

## 0.3.0
* Add test for Portuguese Civ Mod.
* Allow omitting the `opts` argument.
* Rename the useless map virt function table pointer field.

## 0.2.0
* Support for African Kingdoms by passing `{ version: 'african-kingdoms' }`.

## 0.1.0
* Initial release.


test/index.js
const fs = require('fs')
const path = require('path')
const glob = require('glob').sync
const test = require('tape')
const { load } = require('../src')

const files = glob(path.join(__dirname, './dat_files/**/*.dat'))

const special = [
  [/\/hd_empires2_x2_p1\.dat$/, {
    version: 'african-kingdoms'
  }]
]

files.forEach((file) => {
  test(path.relative(__dirname, file), {
    // AoK does not yet work
    skip: /\/empires2\.dat$/.test(file)
  }, (t) => {
    t.plan(3)
    fs.readFile(file, (err, buf) => {
      t.ifError(err)
      const opts = special.reduce((acc, [name, opts]) => {
        if (acc) return acc
        if (name.test(file)) return opts
        return undefined
      }, undefined)

      load(buf, opts, (err, dat) => {
        t.ifError(err)
        t.ok(dat)
      })
    })
  })
})


src/index.js
const { inflateRaw } = require("zlib");
const assign = require("simple-assign");
const structures = require("./struct");
const { DatFile } = structures;

const versions = ["aok", "aoc", "african-kingdoms", "aoe2de", "swgb"];

function check(opts) {
  if (opts.version && versions.indexOf(opts.version) === -1) {
    throw new Error(`genie-dat: invalid version: ${opts.version}`);
  }
}

function load(compressedBuffer, opts = {}, cb) {
  if (typeof opts === "function") {
    cb = opts;
    opts = {};
  }
  check(opts);
  inflateRaw(compressedBuffer, (err, uncompressedBuffer) => {
    if (err) return cb(err);
    // try {
      DatFile.version = opts.version || "aoe2de";
      console.log('DatFile.version:', DatFile.version)
      cb(null, DatFile(uncompressedBuffer));
    // } catch (err) {
      // cb(err);
    // } finally {
      // DatFile.version = "aoe2de";
    // }
  });
}

function loadRaw(uncompressedBuffer, opts = {}, cb) {
  if (typeof opts === "function") {
    cb = opts;
    opts = {};
  }
  check(opts);
  process.nextTick(() => {
    try {
      DatFile.version = opts.version || "aoe2de";
      cb(null, DatFile(uncompressedBuffer));
    } catch (err) {
      cb(err);
    } finally {
      DatFile.version = "aoe2de";
    }
  });
}

module.exports = assign({ load, loadRaw }, structures);


src/versions.js
const struct = require("awestruct");
const t = struct.types;

// Increasing variables to represent an ENUM
const GV_None = 1;
const GV_TEST = 2;
const GV_MIK = 3;
const GV_DAVE = 4;
const GV_MATT = 5;
const GV_AoEB = 6;
const GV_AoE = 7;
const GV_RoR = 8;
const GV_Tapsa = 9;
const GV_T2 = 10;
const GV_T3 = 11;
const GV_T4 = 12;
const GV_T5 = 13;
const GV_T6 = 14;
const GV_T7 = 15;
const GV_T8 = 16;
const GV_AoKE3 = 17;
const GV_AoKA = 18;
const GV_AoKB = 19;
const GV_AoK = 21;
const GV_TC = 22;
const GV_TCV = 23;
const GV_Cysion = 24;
const GV_C2 = 25;
const GV_C3 = 26;
const GV_C4 = 27;
const GV_CK = 28;
const GV_C5 = 29;
const GV_C6 = 30;
const GV_C7 = 31;
const GV_C8 = 32;
const GV_C9 = 33;
const GV_C10 = 34;
const GV_C11 = 35;
const GV_C12 = 36;
const GV_C13 = 37;
const GV_C14 = 38;
const GV_C15 = 39;
const GV_C16 = 40;
const GV_C17 = 41;
const GV_C18 = 42;
const GV_C19 = 43;
const GV_C20 = 44;
const GV_SWGB = 45;
const GV_CC = 46;
const GV_CCV = 47;
const GV_CCV2 = 48;


const GV_LatestDE2 = GV_C20;

const VERSIONS = {
  GV_None: 0, //Game version not set
  GV_TEST, // ?
  GV_MIK, // ?
  GV_DAVE, // ?
  GV_MATT, // ?
  GV_AoEB, // 7.04 - 7.11
  GV_AoE, // 7.2
  GV_RoR, // 7.24
  GV_Tapsa, GV_T2, GV_T3, GV_T4, GV_T5, GV_T6, GV_T7, GV_T8, // 10.1 - 10.8
  GV_AoKE3, // 9.36
  GV_AoKA, // 10.19
  GV_AoKB, // 11.05
  GV_AoK, // 11.5
  GV_TC, // 11.76
  GV_TCV, // Terrain patch
  GV_Cysion, // 12.0
  GV_C2, GV_C3, GV_C4, GV_CK, GV_C5, GV_C6, GV_C7, GV_C8, GV_C9, GV_C10, GV_C11, GV_C12, GV_C13, GV_C14, // 12.52 - 12.94
  GV_C15, // 13.11
  GV_C16, // 20.01
  GV_C17, // 20.14
  GV_C18, // 25.27
  GV_C19, GV_C20, // 26.23 - 40.3
  GV_SWGB, // 1.0
  GV_CC, // 1.1
  GV_CCV, // Terrain patch
  GV_CCV2 // Terrain patch + tech tree patch
}

const VER_MAP = {
  "VER 4.5": GV_T8,
  "VER 7.1": GV_C14,
  "VER 7.2": GV_C15,
  "VER 7.3": GV_C16,
  "VER 7.4": GV_C17,
  "VER 7.5": GV_C18,
  "VER 7.6": GV_C19,
  "VER 7.7": GV_C20,
}

const ifVersion = (test, consequent) =>
  t.if(s => {
    let parent = s;
    while (parent.$parent) parent = parent.$parent;
    const { fileVersion } = parent;

    return test(VER_MAP[fileVersion]);
  }, consequent);

module.exports = {
  VER_MAP,
  ifVersion,
  VERSIONS,
  GV_None, 
  GV_TEST, 
  GV_MIK, 
  GV_DAVE, 
  GV_MATT, 
  GV_AoEB, 
  GV_AoE, 
  GV_RoR, 
  GV_Tapsa, 
  GV_T2, 
  GV_T3, 
  GV_T4, 
  GV_T5, 
  GV_T6, 
  GV_T7, 
  GV_T8, 
  GV_AoKE3, 
  GV_AoKA, 
  GV_AoKB, 
  GV_AoK, 
  GV_TC, 
  GV_TCV, 
  GV_Cysion, 
  GV_C2, 
  GV_C3, 
  GV_C4, 
  GV_CK, 
  GV_C5, 
  GV_C6, 
  GV_C7, 
  GV_C8, 
  GV_C9, 
  GV_C10, 
  GV_C11, 
  GV_C12, 
  GV_C13, 
  GV_C14, 
  GV_C15, 
  GV_C16, 
  GV_C17, 
  GV_C18, 
  GV_C19, 
  GV_C20, 
  GV_SWGB, 
  GV_CC, 
  GV_CCV, 
  GV_CCV2, 
  GV_LatestDE2,
}

src/struct.js
const struct = require("awestruct");
const {
  ResourceStorage,
  DamageGraphic,
  StaticObject,
  AnimatedObject,
  DoppelgangerObject,
  MovingObject,
  ActionObject,
  HitType,
  ProjectileObject,
  MissileObject,
  ObjectCommand,
  ResourceCost,
  LivingObject,
  BuildingAnnex,
  BuildingObject,
  TreeObject,
  TriageObject
} = require("./object");
const TechTree = require("./tech-tree");
const DebugString = require("./string");
const { ifVersion } = require("./versions");
const t = struct.types;

// Struct.types.array but it adds an $index property
const indexArray = (size, elementType) => {
  const typeClass = struct.getType(elementType);
  return struct.Type({
    read(opts, parent) {
      const l = struct.getValue(opts.struct, size);
      const result = [];
      for (let i = 0; i < l; i++) {
        opts.struct.$index = i;
        result.push(typeClass.read(opts, parent));
      }
      delete opts.struct.$index;
      return result;
    },
    write(opts, value) {
      const l = struct.getValue(opts.struct, size);
      if (value.length !== l) {
        throw new Error("cannot write incorrect array length, expected " + l + ", got " + value.length);
      }
      for (let i = 0; i < l; i++) {
        opts.struct.$index = i;
        typeClass.write(opts, value[i]);
      }
      delete opts.struct.$index;
    },
    size: typeof size === "number" ? (value, struct) => size * typeClass.size(value[0], struct) : (value, struct) => (value.length ? typeClass.size(value[0], struct) * value.length : 0)
  });
};

// An array of size `size`, but with missing elements specified by `pointers`.
// If `pointers[i] == 0`, the element is skipped.
const holeyArray = (size, pointers, elementType) =>
  indexArray(
    size,
    t.if(s => {
      const ptr = struct.getValue(s, pointers);
      if (!Array.isArray(ptr)) throw new Error(`${pointers} does not refer to a pointer array`);
      return ptr[s.$index] !== 0;
    }, elementType)
  ).mapRead(
    // Filter null elements
    arr => arr.filter(Boolean)
  );

function getTerrainCount() {
  if (DatFile.version === "swgb") return 55;
  if (DatFile.version === "african-kingdoms") return 100;
  if (DatFile.version === "aoe2de") return 200;
  if (DatFile.version !== "aoc") return 32;
  return 42;
}

const PlayerColor = struct([
  ["id", t.int32],
  ["base", t.int32],
  ["unitOutlineColor", t.int32],
  ["unitSelectionColor1", t.int32],
  ["unitSelectionColor2", t.int32],
  ["minimapColor1", t.int32],
  ["minimapColor2", t.int32],
  ["minimapColor3", t.int32],
  ["statisticsTextColor", t.int32]
]);

const TerrainPassGraphic = struct([
  ["exitTileId", t.int32],
  ["enterTileId", t.int32],
  ["walkTileId", t.int32],
  ifVersion(v => false /* SWGB */, struct([["WalkSpriteRateF", t.float]])).else(struct([["WalkSpriteRate", t.int32]]))
]);

const TerrainRestriction = count =>
  struct([
    ["accessibleDamageMultiplier", t.array(s => struct.getValue(s.$parent, count), t.float)],
    ifVersion(v => true /* AOK and up */, struct([["TerrainPassGraphics", t.array(s => struct.getValue(s.$parent.$parent, count), TerrainPassGraphic)]]))
  ]);

const SoundItem = struct([
  ["filename", DebugString],
  ["resourceId", t.int32],
  ["probability", t.int16],
  ifVersion(
    v => true /* AOK and up */,
    struct([
      ["civilization", t.int16],
      ["iconSet", t.int16]
    ])
  )
]);

const Sound = struct([
  ["id", t.int16],
  ["playDelay", t.int16],
  ["fileCount", t.uint16],
  ["cacheTime", t.int32], // gv>GV_test
  ["totalProbability", t.int16],
  ["items", t.array("fileCount", SoundItem)]
]);

const GraphicDelta = struct([
  ["graphicId", t.int16],
  ["padding1", t.int16],
  ["spritePointer", t.int32],
  ["offsetX", t.int16],
  ["offsetY", t.int16],
  ["displayAngle", t.int16],
  ["padding2", t.int16]
]);

const SoundProp = struct([
  ["delay", t.int16],
  ["id", t.int16],
  ["wwiseSoundId", t.uint32]
]);

const GraphicAttackSound = struct([["soundProps", t.array(3, SoundProp)]]);

const Graphic = struct([
  ["name", DebugString],
  ["filename", DebugString],
  ["particleEffectName", DebugString],
  ["slpId", t.int32],
  ["isLoaded", t.int8],
  ["oldColorFlag", t.int8],
  ["layer", t.int8],
  ["playerColor", t.int8],
  ["adaptColor", t.int8],
  ["transparentSelection", t.int8],
  ["coordinates", t.array(4, t.int16)],
  ["deltaCount", t.uint16],
  ["soundId", t.int16],
  ["wwiseSoundId", t.uint32],
  ["attackSoundUsed", t.int8],
  ["frameCount", t.uint16],
  ["angleCount", t.uint16],
  ["speedAdjust", t.float],
  ["frameRate", t.float],
  ["replayDelay", t.float],
  ["sequenceType", t.int8],
  ["id", t.int16],
  ["mirroringMode", t.int8],
  ifVersion(v => true /* AOK and up */, struct([["editorFlag", t.int8]])),
  ["deltas", t.array("deltaCount", GraphicDelta)],
  t.if("attackSoundUsed", struct([["attackSounds", t.array("../angleCount", GraphicAttackSound)]]))
]);

const TileSize = struct([
  ["width", t.int16],
  ["height", t.int16],
  ["deltaY", t.int16]
]);

const TerrainAnimation = struct([
  ["isAnimated", t.bool],
  ["animationFrameCount", t.int16],
  ["pauseFrameCount", t.int16],
  ["interval", t.float],
  ["pauseBetweenLoops", t.float],
  ["frame", t.int16],
  ["drawFrame", t.int16],
  ["animateLast", t.float],
  ["frameChanged", t.bool],
  ["drawn", t.int8]
]);

const FrameData = struct([
  ["frameCount", t.int16],
  ["angleCount", t.int16],
  ["shapeId", t.int16]
]);

const Terrain = struct([
  ["enabled", t.bool],
  ["random", t.int8],
  ["isWater", t.int8],
  ["hideInEditor", t.int8],
  ["stringId", t.int32],
  ["name0", DebugString],
  ["name1", DebugString],
  ["slpId", t.int32],
  ["shapePointer", t.int32],
  ["soundId", t.int32],
  ["wwiseSoundId", t.uint32],
  ["wwiseSoundStopId", t.uint32],
  ifVersion(
    v => true /* AOK and up */,
    struct([
      ["blendPriority", t.int32],
      ["blendMode", t.int32]
    ])
  ),
  ["overlayMaskName", DebugString],
  [
    "minimap",
    struct([
      ["high", t.uint8],
      ["medium", t.uint8],
      ["low", t.uint8],
      ["cliffLt", t.uint8],
      ["cliffRt", t.uint8]
    ])
  ],
  ["passableTerrain", t.int8],
  ["impassableTerrain", t.int8],
  TerrainAnimation,
  ["elevationGraphics", t.array(19, FrameData)],
  ["terrainReplacementId", t.int16],
  ["rows", t.int16],
  ["columns", t.int16],
  ["terrainUnitMaskedDensity", t.array(30, t.int16)],
  ["terrainUnitId", t.array(30, t.int16)],
  ["terrainUnitDensity", t.array(30, t.int16)],
  ["terrainPlacementFlag", t.array(30, t.int8)],
  ["terrainUnitsUsedCount", t.int16],
  ifVersion(v => true /* not SWGB */, struct([["phantom", t.int16]]))
]);

const TerrainBorder = struct([
  ["enabled", t.bool],
  ["random", t.int8],
  ["name0", t.char(13)],
  ["name1", t.char(13)],
  ["slpId", t.int32],
  ["shapePointer", t.int32],
  ["soundId", t.int32],
  ["color", t.array(3, t.uint8)],
  TerrainAnimation,
  ["frames", t.array(19 * 12, FrameData)],
  ["drawTile", t.int16],
  ["underlayTerrain", t.int16],
  ["borderStyle", t.int16]
]);


const RandomMapLand = struct([
  ["landId", t.int32],
  ["terrain", t.int32],
  ["spacing", t.int32],
  ["baseSize", t.int32],
  ["zone", t.int8],
  ["placementType", t.int8],
  t.skip(2), // padding
  ["baseX", t.int32],
  ["baseY", t.int32],
  ["landSize", t.int8],
  ["byPlayerFlag", t.int8],
  t.skip(2), // padding
  ["startAreaRadius", t.int32],
  ["borderFuzziness", t.int32],
  ["clumpiness", t.int32]
]);

const RandomMapTerrain = struct([
  ["size", t.int32],
  ["terrain", t.int32],
  ["numberOfClumps", t.int32],
  ["edgeSpacing", t.int32],
  ["placementZone", t.int32],
  ["clumpiness", t.int32]
]);

const RandomMapObject = struct([
  ["objectId", t.int32],
  ["baseTerrain", t.int32],
  ["groupingType", t.int8],
  ["scalingType", t.int8],
  t.skip(2), // padding
  ["objectsPerGroup", t.int32],
  ["groupVariance", t.int32],
  ["numberOfGroups", t.int32],
  ["groupRadius", t.int32],
  ["ownAtStart", t.int32],
  ["setPlaceForAllPlayers", t.int32],
  ["minDistanceToPlayers", t.int32],
  ["maxDistanceToPlayers", t.int32]
]);

const RandomMapElevation = struct([
  ["size", t.int32],
  ["terrain", t.int32],
  ["numberOfClumps", t.int32],
  ["baseTerrain", t.int32],
  ["baseElevation", t.int32],
  ["tileSpacing", t.int32]
]);

const RandomMapInfo = struct([
  ["mapId", t.int32],
  ["borderLeft", t.int32],
  ["borderTop", t.int32],
  ["borderRight", t.int32],
  ["borderBottom", t.int32],
  ["borderUsage", t.int32],
  ["waterShape", t.int32],
  ["baseTerrain", t.int32],
  ["landCoverage", t.int32],
  ["unusedId", t.int32],
  ["baseZoneCount", t.uint32],
  ["baseZonePointer", t.int32],
  ["terrainCount", t.uint32],
  ["terrainPointer", t.int32],
  ["objectCount", t.uint32],
  ["objectPointer", t.int32],
  ["elevationCount", t.uint32],
  ["elevationPointer", t.int32]
]);
const RandomMap = struct([
  ["borderLeft", t.int32],
  ["borderTop", t.int32],
  ["borderRight", t.int32],
  ["borderBottom", t.int32],
  ["borderUsage", t.int32],
  ["waterShape", t.int32],
  ["baseTerrain", t.int32],
  ["landCoverage", t.int32],
  ["unusedId", t.int32],
  ["baseZoneCount", t.uint32],
  ["baseZonePointer", t.int32],
  ["baseZones", t.array("baseZoneCount", RandomMapLand)],
  ["terrainCount", t.uint32],
  ["terrainPointer", t.int32],
  ["terrains", t.array("terrainCount", RandomMapTerrain)],
  ["objectCount", t.uint32],
  ["objectPointer", t.int32],
  ["objects", t.array("objectCount", RandomMapObject)],
  ["elevationCount", t.uint32],
  ["elevationPointer", t.int32],
  ["elevations", t.array("elevationCount", RandomMapElevation)]
]);

const TechEffect = struct([
  ["type", t.int8],
  ["unit", t.int16],
  ["unitClassId", t.int16],
  ["attributeId", t.int16],
  ["amount", t.float]
]);

const Tech = struct([
  ["name", DebugString],
  ["effects", t.dynarray(t.uint16, TechEffect)]
]);

const ObjectHeader = struct([["exists", t.bool], t.if("exists", struct([["commands", t.dynarray(t.uint16, ObjectCommand)]]))]);

const Civilization = struct([
  ["playerType", t.int8],
  ["name", DebugString],
  ["resourcesCount", t.uint16],
  ["techTreeId", t.int16],
  ["teamBonusId", t.int16], // not in AOE1
  ["resources", t.array("resourcesCount", t.float)],
  ["iconSet", t.int8],
  ["unitCount", t.uint16],
  ["unitPtrs", t.array("unitCount", t.int32)],
  ["objects", holeyArray("unitCount", "unitPtrs", TriageObject)]
]);

const ResearchResourceCost = struct([
  ["resourceId", t.int16],
  ["amount", t.int16],
  ["enabled", t.bool]
]);

const Research = struct([
  ["requiredTechIds", t.array(6, t.int16)],
  ["resourceCosts", t.array(3, ResearchResourceCost)],
  ["requiredTechCount", t.int16],
  ["civilizationId", t.int16],
  ["fullTechMode", t.int16],
  ["researchLocationId", t.int16],
  ["languageDllName", t.int32], // >= GV_C18
  ["languageDllDescription", t.uint32],
  ["researchTime", t.int16],
  ["techEffectId", t.int16],
  ["techType", t.int16],
  ["iconId", t.int16],
  ["buttonId", t.int8],
  ["languageDllHelp", t.int32],
  ["languageDllTechTree", t.int32],
  ["hotkey", t.int32],
  ["name", DebugString],
  ["repeatable", t.int8],
  // SWGB
  // ['name2', t.dynstring(t.uint16)]
]);

const DatFile = struct([
  ["fileVersion", t.char(8), val => console.error(val)],
  ["terrainRestrictionCount", t.uint16],
  // ["SUnknown2", t.uint32],
  // ["SUnknown3", t.uint32],
  // ["SUnknown4", t.uint32],
  // ["SUnknown5", t.uint32],
  ["terrainCount", t.uint16],
  ["terrainTables", t.array("terrainRestrictionCount", t.int32)],
  ["terrainPassGraphicPointers", t.array("terrainRestrictionCount", t.int32)],
  ["terrainRestrictions", t.array("terrainRestrictionCount", TerrainRestriction("terrainCount"))],
  ["playerColors", t.dynarray(t.uint16, PlayerColor)],
  ["soundsCount", t.uint16],
  // ["sounds", t.dynarray(t.uint16, Sound)],
  ["sounds", t.array('soundsCount', Sound), console.warn],
  ["graphicCount", t.uint16],
  ["graphicPtrs", t.array("graphicCount", t.int32)],
  ["graphics", holeyArray("graphicCount", "graphicPtrs", Graphic)],
  ["RGE_Map::vfptr", t.int32],
  ["mapPointer", t.int32],
  ["mapWidth", t.int32],
  ["mapHeight", t.int32],
  ["worldWidth", t.int32],
  ["worldHeight", t.int32],
  ["tileSizes", t.array(19, TileSize)],
  ["paddingTS1", t.int16],
  ["terrains", t.array(getTerrainCount, Terrain)],
  ["mapMinX", t.float],
  ["mapMinY", t.float],
  ["mapMaxX", t.float],
  ["mapMaxY", t.float],
  ["mapMaxX+1", t.float],
  ["mapMaxY+1", t.float],
  ["terrainsUsed2", t.uint16], // additionalTerrainCount oldname?
  ["bordersUsed", t.uint16],
  ["maxTerrain", t.int16],
  ["tileWidth", t.int16],
  ["tileHeight", t.int16],
  ["tileHalfHeight", t.int16],
  ["tileHalfWidth", t.int16],
  ["elevHeight", t.int16],
  ["currentRow", t.int16],
  ["currentColumn", t.int16],
  ["blockBeginRow", t.int16],
  ["blockEndRow", t.int16],
  ["blockBeginColumn", t.int16],
  ["blockEndColumn", t.int16],
  ["searchMapPointer", t.int32],
  ["searchMapRowsPointer", t.int32],
  ["anyFrameChange", t.int8],
  ["mapVisibleFlag", t.int8],
  ["fogFlag", t.int8], // Always 1
  ["randomMapCount", t.uint32],
  ["randomMapPointer", t.int32],
  ["randomMapInfo", t.array("randomMapCount", RandomMapInfo)],
  ["randomMaps", t.array("randomMapCount", RandomMap)],

  ["effects", t.dynarray(t.uint32, Tech)],
  // TODO Unit lines for SWGB
  ["objectHeaders", t.dynarray(t.uint32, ObjectHeader)],
  ["civilizations", t.dynarray(t.uint16, Civilization)],
  ["researches", t.dynarray(t.uint16, Research)],
  ["timeSlice", t.int32],
  ["unitKillRate", t.int32],
  ["unitKillTotal", t.int32],
  ["unitHitPointRate", t.int32],
  ["unitHitPointTotal", t.int32],
  ["razingKillRate", t.int32],
  ["razingKillTotal", t.int32],
  ["techTree", TechTree]
]);

module.exports = {
  PlayerColor,
  TerrainPassGraphic,
  SoundItem,
  Sound,
  GraphicDelta,
  SoundProp,
  GraphicAttackSound,
  Graphic,
  TileSize,
  TerrainAnimation,
  FrameData,
  Terrain,
  TerrainBorder,
  RandomMapInfo,
  RandomMapLand,
  RandomMapTerrain,
  RandomMapObject,
  RandomMapElevation,
  RandomMap,
  TechEffect,
  Tech,
  ObjectHeader,
  ResourceStorage,
  DamageGraphic,
  StaticObject,
  AnimatedObject,
  DoppelgangerObject,
  MovingObject,
  ActionObject,
  HitType,
  ProjectileObject,
  MissileObject,
  ResourceCost,
  LivingObject,
  BuildingAnnex,
  BuildingObject,
  TreeObject,
  TriageObject,
  Civilization,
  ResearchResourceCost,
  Research,
  DatFile
};
